src/api/base.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  Status,
+  ModuleProfile,
+  Api,
+  PluginProfile,
+  ApiEventEmitter,
+  PluginApi,
+  PluginRequest,
+  ExtractKey,
+} from '../types'
+import { EventEmitter } from 'events'
+
+type Profile<T extends Api> = ModuleProfile<T> | PluginProfile<T>
+
+/**
+ * Create a Profile with default values
+ * @param profile The profile of the plugin
+ * @param mixinProfiles A list of profiles to mix in the base profile
+ */
+export function createProfile<T extends Api, U extends Api>(
+  profile: Profile<T>,
+  ...mixinProfiles: Partial<ModuleProfile<U>>[]
+): Profile<T> {
+  return {
+    ...profile,
+    methods: [
+      ...mixinProfiles.reduce((acc, mixin) => [ ...acc, ...(mixin.methods || []) ], []),
+      ...(profile.methods || [])
+    ],
+    events: [
+      ...mixinProfiles.reduce((acc, mixin) => [ ...acc, ...(mixin.events || []) ], []),
+      ...(profile.events || [])
+    ],
+    notifications: {
+      ...mixinProfiles.reduce((acc, mixin) => ({ ...acc, ...(mixin.notifications || {}) }), {} as any),
+      ...(profile.notifications || {}),
+    },
+  }
+}
+
+export const baseProfile = {
+  notifications: {
+    'theme': ['switchTheme']
+  }
+}
+
+export abstract class BaseApi<S, U extends Api> {
+  private initialState: S
+  protected state: S
```

Comment:
why do you keep `initialState` here?

Reply:
nevermind
https://github.com/ethereum/remix-plugin/pull/58/files#diff-74ebb899b68ffa95e80a782b8980aa1bR96